README.md
Nimrod argument parser
======================

[Nimrod](http://nimrod-code.org) provides the [parseopt
module](http://nimrod-code.org/parseopt.html) to parse options from
the command line. I found this module lacking, used to python modules
like [optparse](http://docs.python.org/2/library/optparse.html) or
[argparse](http://docs.python.org/3/library/argparse.html).  This
module tries to provide similar functionality to prevent you from
writing command line parsing and let you concentrate on providing
the best possible experience for your users.


License
=======

[MIT license](LICENSE.md).


Installation and usage
======================

To get the source code you can likely use [Nimrod's babel package
manager](https://github.com/nimrod-code/babel) and type:

    babel install argument_parser

If you don't have babel you can simply download the argument_parser.nim
file and add it to your program. All other files in this repository
are accessory and can be ignored for normal usage. Once you have
the source you can just ``import argument_parser`` and use its
contents. Use [nimrod's configuration
files](http://nimrod-code.org/nimrodc.html#configuration-files)
feature to specify a path to where you install this module so you
don't have to copy it around.


Documentation
=============

The argument_parser module comes with embedded docstrings. You can
run ``nimrod doc2 argument_parser.nim`` and obtain a reference html
file with instructions on the exported symbols.  If you installed
through babel, you can find this in a path similar to
``~/.babel/libs/argument_parser-version``.

The generated documentation for all public versions can also be
found at
[http://gradha.github.io/argument_parser/](http://gradha.github.io/argument_parser/).
No guarantees on its freshness, though, do check the generation
date at the bottom.

In the distant future, when most features are complete, a tutorial
will be provided to explain how to use the module. In the meantime
you should read the examples provided in the [examples
subdirectory](examples). These examples try to show how to implement
common usage patterns for different types of command line parsing.


Changes
=======

The last version is 0.3.0, see the [list of recent changes](CHANGES.md).


Git branches
============

This project uses the `git-flow branching model
<https://github.com/nvie/gitflow>`_. Which means the ``master`` default branch
doesn't *see* much movement, development happens in another branch like
``develop``. Most people will be fine using the ``master`` branch, but if you
want to contribute something please check out first the ``develop`` branch and
do pull requests against that.


Feedback
========

You can send me feedback through [github's issue
tracker](http://github.com/gradha/argument_parser/issues). I also
take a look from time to time to [Nimrod's
forums](http://forum.nimrod-code.org) where you can talk to other
nimrod programmers.


CHANGES.md
Changes
=======

v0.3.0 ????-??-??
-----------------

* [Adopted git-flow](https://github.com/gradha/argument_parser/issues/25).

v0.1.1 2013-04-13
-----------------

* Fixes testing for options with various names
  ([#20](https://github.com/gradha/argument_parser/issues/20)).

v0.1.0 2013-03-10
-----------------

* Initial release.


tests/example1.nim
import argument_parser, strutils, tables

proc process_commandline(): Tcommandline_results =
  let
    p1 = new_parameter_specification(pkString,
      help_text = "Three weird params", names = @["-alt", "-a", "--arg"])
    p2 = new_parameter_specification(pkString,
      help_text = "This one felt so lonely", names = "--aasd")
    p3 = new_parameter_specification(pkInt,
      help_text = "Integers for the win!", names = "-i")
    p4 = new_parameter_specification(pkFloat,
      help_text = "Meh, floats are superior", names = @["-f", "--floatts"])
    p5 = new_parameter_specification(pkBool,
      help_text = "The next parameter always lies", names = "-b")
    p6 = new_parameter_specification(pkBiggestInt,
      help_text = "Admire my size", names = "-I")
    p7 = new_parameter_specification(pkBiggestFloat,
      help_text = "Admire my precission", names = "-F")
    p8 = new_parameter_specification(help_text = "Shows version and exits",
      names = @["-v", "--verbose"])
    ph = new_parameter_specification(pkHelp,
      help_text = "Display help and quit", names = @["-h", "--help", "help"])
    all_params = @[p1, p2, p3, p4, p5, p6, p7, p8, ph]
  result = parse(all_params)

when isMainModule:
  let ret = process_commandline()
  echo($ret)


tests/code_test.nim
import argument_parser, strutils, tables

template test_failure(exception, code: expr): stmt =
  try:
    let pos = instantiationInfo()
    discard(code)
    echo "Unit test failure at $1:$2 with '$3'" % [pos.filename,
      $pos.line, astToStr(code)]
    doAssert false, "A test expecting failure succeeded?"
  except exception:
    nil

#template test_success(code: expr) {.immediate.} = echo($code)
template test_success(code: expr) {.immediate.} = discard(code)

template test_in(commandline_parameters, expected_value, attribute: expr):
    expr {.immediate.} =
  ## Returns true if the expected value is found in the commandline parameters.
  ##
  ## Pass the attribute to check in the Tparsed_parameter object.
  var result {.gensym.}: bool = false
  for parameter in commandline_parameters.positional_parameters:
    if parameter.attribute == expected_value:
      result = true
      break
  result

template tp(expected: seq[Tparameter_specification] = @[],
    type_of_positional_parameters = PK_STRING, args: seq[TaintedString] = nil,
    bad_prefixes = @["-", "--"], end_of_options = "--"):
      expr =
  # Simple wrapper to avoid changing the last default parameter.
  parse(expected, type_of_positional_parameters, args,
    bad_prefixes, end_of_options, quit_on_failure = false)

proc test() =
  #echo "\nParsing default system params"
  let
    p1 = new_parameter_specification(pkString, names = "-a")
    p2 = new_parameter_specification(pkString, names = "--aasd")
    p3 = new_parameter_specification(pkInt, names = "-i")
    p4 = new_parameter_specification(pkFloat, names = "-f")
    p5 = new_parameter_specification(pkBool, names = "-b")
    p6 = new_parameter_specification(pkBiggestInt, names = "-I")
    p7 = new_parameter_specification(pkBiggestFloat, names = "-F")
    all_params = @[p1, p2, p3, p4, p5, p6, p7]

  discard(tp(all_params))

  let args = @["test", "toca me la", "-a", "-wo", "rd", "--aasd", "--s", "ugh"]
  #echo "\nParsing ", join(args, " ")
  var ret = tp(all_params, args = args)
  #echo($ret)
  doAssert ret.options["-a"].strVal == "-wo"
  doAssert ret.options.hasKey("test") == false
  doAssert test_in(ret, "test", str_val)
  doAssert test_in(ret, "--s", str_val) == false
  doAssert test_in(ret, "ugh", str_val)
  test_failure(EInvalidValue, tp(all_params, PK_INT, args))

  # Integer tests.
  test_success(tp(all_params, args = @["int test", "-i", "445"]))
  test_success(tp(all_params, args = @["int test", "-i", "-445"]))
  test_failure(EInvalidValue, tp(all_params, args = @["-i", ""]))
  test_failure(EInvalidValue, tp(all_params, args = @["-i", "0x02"]))
  test_failure(EInvalidValue, tp(all_params, args = @["-i", "fail"]))
  test_failure(EInvalidValue, tp(all_params, args = @["-i", "234.12"]))
  test_failure(EOverflow, tp(all_params, args = @["-i", $high(int) & "0"]))
  test_success(tp(all_params, PK_INT, @["-i", "-445", "2", "3", "4"]))
  test_failure(EInvalidValue,
    tp(all_params, PK_INT, @["-i", "-445", "2", "3", "4.3"]))

  ret = tp(all_params, PK_INT, @["-i", "-445", "2", "3", "4"])
  doAssert ret.options["-i"].int_val == -445
  doAssert test_in(ret, 2, int_val)
  doAssert test_in(ret, 3, int_val)
  doAssert test_in(ret, 4, int_val)
  doAssert test_in(ret, 5, int_val) == false

  # String tests.
  test_success(tp(all_params, args = @["str test", "-a", "word"]))
  test_success(tp(all_params, args = @["str empty test", "-a", ""]))
  test_failure(EInvalidValue, tp(all_params, args = @["str test", "-a"]))

  # Float tests.
  test_success(tp(all_params, args = @["-f", "123.235"]))
  test_failure(EInvalidValue, tp(all_params, args = @["-f", ""]))
  test_failure(EInvalidValue, tp(all_params, args = @["-f", "abracadabra"]))
  test_failure(EInvalidValue, tp(all_params, args = @["-f", "12.34aadd"]))
  ret = tp(all_params, PK_FLOAT, @["-f", "12.23", "89.2", "3.14"])
  doAssert ret.options["-f"].float_val == 12.23
  doAssert test_in(ret, 89.2, float_val)
  doAssert test_in(ret, 3.14, float_val)
  doAssert test_in(ret, 3.1, float_val) == false

  # Boolean tests.
  for param in @["y", "yes", "true", "1", "on", "n", "no", "false", "0", "off"]:
    test_success(tp(all_params, args = @["-b", param]))
  test_failure(EInvalidValue, tp(all_params, args = @["-b", "t"]))
  test_failure(EInvalidValue, tp(all_params, args = @["-b", ""]))
  doAssert test_in(tp(all_params, PK_BOOL, @["y"]), true, bool_val)
  doAssert test_in(tp(all_params, PK_BOOL, @["0"]), false, bool_val)

  # Big integer tests.
  test_success(tp(all_params, args = @["int test", "-I", "445"]))
  test_failure(EInvalidValue, tp(all_params, args = @["-I", ""]))
  test_failure(EInvalidValue, tp(all_params, args = @["-I", "fail"]))
  test_failure(EInvalidValue, tp(all_params, args = @["-I", "234.12"]))
  test_failure(EOverflow, tp(all_params,
    args = @["-I", $high(biggestInt) & "0"]))
  ret = tp(all_params, PK_BIGGEST_INT, @["42", $high(biggestInt)])
  doAssert test_in(ret, 42, big_int_val)
  doAssert test_in(ret, high(biggestInt), big_int_val)
  doAssert test_in(ret, 13, big_int_val) == false

  # Big float tests.
  test_success(tp(all_params, args = @["-F", "123.235"]))
  test_failure(EInvalidValue, tp(all_params, args = @["-F", ""]))
  test_failure(EInvalidValue, tp(all_params, args = @["-F", "abracadabra"]))
  test_failure(EInvalidValue, tp(all_params, args = @["-F", "12.34aadd"]))
  ret = tp(all_params, PK_BIGGEST_FLOAT, @["111.111", "9.01"])
  doAssert test_in(ret, 111.111, big_float_val)
  doAssert test_in(ret, 9.01, big_float_val)
  doAssert test_in(ret, 9.02, big_float_val) == false

  # Using custom procs for transformation of type back to string.
  var c1 = new_parameter_specification(PK_INT, names = "-i")
  ret = tp(@[c1], args = @["-i", "42"])
  doAssert ret.options["-i"].int_val == 42
  # Now repeat transforming to int through a custom post-validator.
  c1.custom_validator =
    proc (parameter: string, value: var Tparsed_parameter): string =
      value = new_parsed_parameter(PK_STRING, $value.int_val)
  ret = tp(@[c1], args = @["-i", "42"])
  doAssert ret.options["-i"].str_val == "42"

  # Change the custom proc to reject values lower than 18.
  c1.custom_validator =
    proc (parameter: string, value: var Tparsed_parameter): string =
      let age = value.int_val
      if age < 18:
        result = "Can't accept minors ($1) parsing arguments" % [$age]
      else:
        value = new_parsed_parameter(PK_STRING, "valid_$1" % [$age])
  ret = tp(@[c1], args = @["-i", "42"])
  doAssert ret.options["-i"].str_val == "valid_42"
  test_failure(EInvalidValue, tp(@[c1], args = @["-i", "17"]))

  # Make sure we disallow multiple parameters being the same.
  test_failure(EInvalidKey,
    tp(@[new_parameter_specification(names = ["-a", "-a"])]))
  # Also repeated keys in different parameters.
  test_failure(EInvalidKey,
    tp(@[new_parameter_specification(names = ["-a", "--alt"]),
      new_parameter_specification(names = ["-p", "--pert"]),
      new_parameter_specification(names = ["-z", "--alt"])]))
  # The following will fail with ambiguos parameter.
  test_failure(EInvalidValue,
    tp(all_params, args = @["-bleah", "something"]))
  # This one won't because we are passing the disambiguator.
  discard(tp(all_params, args = @["--", "-bleah", "something"]))
  # This should work too because we changed the bad prefixes.
  discard(tp(all_params, args = @["-bl", "so"], bad_prefixes = @[]))
  # This should detect new prefixes.
  test_failure(EInvalidValue,
    tp(all_params, args = @["/bl", "so"], bad_prefixes = @["/"]))
  # Mix new prefixes plus end of parsing options.
  discard(tp(all_params, args = @["-*-", "/bĸ", "a"],
    bad_prefixes = @["/"], end_of_options = "-*-"))

  # Test boolean switches using their second version.
  ret = tp(@[new_parameter_specification(names = @["-s", "--silent"])],
    args = @["file1", "--silent"])
  doAssert ret.options.hasKey("-s")
  doAssert (not ret.options.hasKey("--silent"))

  ret = tp(@[new_parameter_specification(names = @["--silent", "-s"])],
    args = @["file1", "--silent"])
  doAssert ret.options.hasKey("--silent")
  doAssert (not ret.options.hasKey("-s"))

  echo "Tester finished"

when isMainModule:
  test()


nakefile.nim
import nake, os, osproc, htmlparser, xmltree, strtabs, times,
  argument_parser, zipfiles

const name = "argument_parser"
let
  modules = @[name]
  rst_files = @["docs"/"changes", "docs"/"release_steps",
    "LICENSE", "README", "docindex"]

proc change_rst_links_to_html(html_file: string) =
  ## Opens the file, iterates hrefs and changes them to .html if they are .rst.
  let html = loadHTML(html_file)
  var DID_CHANGE: bool

  for a in html.findAll("a"):
    let href = a.attrs["href"]
    if not href.isNil:
      let (dir, filename, ext) = splitFile(href)
      if cmpIgnoreCase(ext, ".rst") == 0:
        a.attrs["href"] = dir / filename & ".html"
        DID_CHANGE = true

  if DID_CHANGE:
    writeFile(html_file, $html)

proc needs_refresh(target: string, src: varargs[string]): bool =
  assert len(src) > 0, "Pass some parameters to check for"
  var TARGET_TIME: float
  try:
    TARGET_TIME = toSeconds(getLastModificationTime(target))
  except EOS:
    return true

  for s in src:
    let srcTime = toSeconds(getLastModificationTime(s))
    if srcTime > TARGET_TIME:
      return true

proc nim_to_rst(nim_file, rst_file: string) =
  ## Reads nim_file and creates into rst_file a *blocked* nim version for HTML.
  let
    name = nim_file.splitFile.name
    title_symbols = repeatChar(name.len, '=')
  var source = "$1\n$2\n$1\n.. code-block:: nimrod\n  " % [title_symbols, name]
  source.add(readFile(nim_file).replace("\n", "\n  "))
  writeFile(rst_file, source)


iterator all_rst_files(): tuple[src, dest: string] =
  for rst_name in rst_files:
    var R: tuple[src, dest: string]
    R.src = rst_name & ".rst"
    # Ignore files if they don't exist, babel version misses some.
    if not R.src.existsFile:
      echo "Ignoring missing ", R.src
      continue
    R.dest = rst_name & ".html"
    yield R

iterator all_examples(): tuple[src, dest: string] =
  # Generates .nim/.html pairs from source in the examples directory.
  for path in "examples".walkDirRec:
    if path.splitFile().ext != ".nim": continue
    var R: tuple[src, dest: string]
    R.src = path
    R.dest = path.changeFileExt("html")
    yield R


task "babel", "Uses babel to install " & name & " locally":
  direshell("babel install -y")
  echo "Installed."

task "doc", "Generates HTML version of the documentation":
  # Generate documentation for the nim modules.
  for module in modules:
    let
      nim_file = module & ".nim"
      html_file = module & ".html"
    if not html_file.needs_refresh(nim_file): continue
    if not shell("nimrod doc --verbosity:0", module):
      quit("Could not generate html doc for " & module)
    else:
      echo "Generated " & html_file

  # Generate html files from the example sources.
  for nim_file, html_file in all_examples():
    if not html_file.needs_refresh(nim_file): continue
    # Create temporary rst file.
    let rst_file = nim_file.changeFileExt("rst")
    nim_to_rst(nim_file, rst_file)
    if not shell("nimrod rst2html --verbosity:0", rst_file):
      quit("Could not generate html doc for " & rst_file)
    else:
      change_rst_links_to_html(html_file)
      echo rst_file & " -> " & html_file
    rst_file.removeFile

  # Generate html files from the rst docs.
  for rst_file, html_file in all_rst_files():
    if not html_file.needs_refresh(rst_file): continue
    if not shell("nimrod rst2html --verbosity:0", rst_file):
      quit("Could not generate html doc for " & rst_file)
    else:
      change_rst_links_to_html(html_file)
      echo rst_file & " -> " & html_file
  echo "All done"

task "check_doc", "Validates rst format for a subset of documentation":
  for rst_file, html_file in all_rst_files():
    echo "Testing ", rst_file
    let (output, exit) = execCmdEx("rst2html.py " & rst_file & " > /dev/null")
    if output.len > 0 or exit != 0:
      echo "Failed python processing of " & rst_file
      echo output

task "clean", "Removes temporal files, mainly":
  removeDir("nimcache")
  for rst_file, html_file in all_rst_files():
    echo "Removing ", html_file
    html_file.removeFile
  for nim_file, html_file in all_examples():
    echo "Removing ", html_file
    html_file.removeFile

task "dist_doc", "Generate zip with documentation":
    runTask "clean"
    runTask "doc"
    let
      dname = name & "-" & VERSION_STR & "-docs"
      zname = dname & ".zip"
    removeFile(name)
    removeFile(zname)
    var Z: TZipArchive
    proc zadd(dest, src: string) =
      echo "Adding '" & src & "' -> '" & dest & "'"
      Z.addFile(dest, src)
    if not Z.open(zname, fmWrite):
      quit("Couldn't open zip " & zname)
    try:
      echo "OPening ", name
      zadd(dname / name & ".html", name & ".html")
      echo "OPening ", name
      for rst_file, html_file in all_rst_files():
        zadd(dname / html_file, html_file)
      echo "OPening ", name
      for nim_file, html_file in all_examples():
        zadd(dname / html_file, html_file)
    finally:
      Z.close
    echo "Built ", zname, " sized ", zname.getFileSize, " bytes."


examples/ex_cp.nim
import argument_parser, tables, strutils, parseutils

## Example defining a simple copy command line program.

const
  PARAM_PRESERVE_ALL = "-a"
  PARAM_FORCE = "-f"
  PARAM_FOLLOW_SOME_SYMBOLIC_LINKS = "-H"
  PARAM_INTERACTIVE = "-i"
  PARAM_FOLLOW_ALL_SYMBOLIC_LINKS = "-L"
  PARAM_NO_OVERWRITE = "-n"
  PARAM_FOLLOW_NO_SYMBOLIC_LINKS = "-P"
  PARAM_PRESERVE_ATTRIBUTES = "-p"
  PARAM_RECURSIVE = "-R"
  PARAM_VERBOSE = "-v"
  PARAM_HELP = "-h"


template P(tnames: varargs[string], thelp: string) =
  ## Helper to avoid repetition of parameter adding boilerplate.
  params.add(new_parameter_specification(PK_EMPTY,
    help_text = thelp, names = tnames))


template got(param: string) =
  ## Just dump the detected options on output.
  if result.options.hasKey(param): echo("Found option '$1'." % [param])


proc process_commandline(): Tcommandline_results =
  ## Parses the commandline.
  ##
  ## Returns a Tcommandline_results with at least two positional parameter,
  ## where the last parameter is implied to be the destination of the copying.
  var params: seq[Tparameter_specification] = @[]

  P(PARAM_PRESERVE_ALL, ("Same as $1 $2 $3 options, preserves " &
    "structure and attributes of files but not directory structure") % [
    PARAM_PRESERVE_ATTRIBUTES, PARAM_FOLLOW_NO_SYMBOLIC_LINKS, PARAM_RECURSIVE])
  P(PARAM_FORCE, "Force overwrite destination files")
  P(PARAM_FOLLOW_SOME_SYMBOLIC_LINKS,
    "Follow symbolic links on the command line")
  P(PARAM_INTERACTIVE, "Prompt before overwritting destination")
  P(PARAM_FOLLOW_ALL_SYMBOLIC_LINKS, "Follow all symbolic links recursively")
  P(PARAM_NO_OVERWRITE, "Do not overwrite destination")
  P(PARAM_FOLLOW_NO_SYMBOLIC_LINKS, "No symbolic links are followed")
  P(PARAM_PRESERVE_ATTRIBUTES, "Attributes are preserved to destination")
  P(PARAM_RECURSIVE, "Follow source directories recursively")
  P(PARAM_VERBOSE, "Be verbose about actions")

  params.add(new_parameter_specification(PK_HELP,
    help_text = "Shows this help on the commandline", names = PARAM_HELP))

  result = parse(params)

  if result.positional_parameters.len < 2:
    echo "Missing parameters, you need to pass the source and dest targets."
    echo_help(params)
    quit()

  got(PARAM_PRESERVE_ALL)
  got(PARAM_FORCE)
  got(PARAM_FOLLOW_SOME_SYMBOLIC_LINKS)
  got(PARAM_INTERACTIVE)
  got(PARAM_FOLLOW_ALL_SYMBOLIC_LINKS)
  got(PARAM_NO_OVERWRITE)
  got(PARAM_FOLLOW_NO_SYMBOLIC_LINKS)
  got(PARAM_PRESERVE_ATTRIBUTES)
  got(PARAM_RECURSIVE)
  got(PARAM_VERBOSE)


when isMainModule:
  let args = process_commandline()
  let dest = args.positional_parameters[args.positional_parameters.len - 1]
  for i in 0..args.positional_parameters.len - 2:
    echo "Copying $1 -> $2" % [args.positional_parameters[i].str_val,
      dest.str_val]


examples/ex_git.nim
import argument_parser, tables, strutils, parseutils

## Example defining a subset of git's functionality
##
## The point of this example is to show that you are not required to use
## options with dashes, you can use normal words too and mix them.
##
## At the moment, however, the parser is still too simple and doesn't handle
## branched options which are required for a true git commandline. That is, all
## the options you specify apply at the same time, so you can't exclude one
## switch being available only for a specific command.
##
## This is planned to be improved in the future both through branched switches
## (https://github.com/gradha/argument_parser/issues/18) and by opening the
## parsing state machine for custom procs which will allow any kind of
## behaviour you want (https://github.com/gradha/argument_parser/issues/19).
##
## Also, the look of the generated help is really bad with mixed commands and
## options. Custom help output will be allowed.

const
  PARAM_VERSION = @["--version"]
  PARAM_HELP = @["--help"]
  PARAM_CONFIG = @["-c"]
  PARAM_PAGINATE = @["-p", "--paginate"]
  COMMAND_ADD = "add"
  COMMAND_MV = "mv"
  COMMAND_RM = "rm"
  COMMAND_DIFF = "diff"
  COMMANDS = @[COMMAND_ADD, COMMAND_MV, COMMAND_RM, COMMAND_DIFF]


type
  Tgit_commandline_results = object of Tcommandline_results ## \
    ## Inherits the normal object to add more fields for convenience.
    command: string


template P(tnames: varargs[string], thelp: string, ttype = PK_EMPTY,
    tcallback: Tparameter_callback = nil) =
  ## Helper to avoid repetition of parameter adding boilerplate.
  params.add(new_parameter_specification(ttype, custom_validator = tcallback,
    help_text = thelp, names = tnames))


template got(param: varargs[string]) =
  ## Just dump the detected options on output.
  if result.options.hasKey(param[0]): echo("Found option '$1'." % [param[0]])


proc parse_configuration_parameter(parameter: string;
    value: var Tparsed_parameter): string =
  ## Simple validator of configuration parameters for PARAM_CONFIG.
  ##
  ## If the user specifies the PARAM_CONFIG option this proc will be called
  ## so we can validate the input. The proc returns a non empty string if
  ## something went wrong with the description of the error, otherwise
  ## execution goes ahead.
  ##
  ## We won't actually parse the validity of the values, but we can try to make
  ## sure there are at least two values separated by an equal sign.
  ##
  ## This validator only accepts values without changing the final output.
  let chunks = split(value.str_val, '=')
  if chunks.len != 2:
    return "Bad syntax for configuration parameter, use x=y"

  for chunk in chunks:
    if chunk.len < 1:
      return "Bad syntax for configuration parameter, use x=y"

  echo "Found valid configuration parameter %1" % [value.str_val]


proc process_commandline(): Tgit_commandline_results =
  ## Parses the commandline.
  ##
  ## Returns a Tcommandline_results with at least two positional parameter,
  ## where the last parameter is implied to be the destination of the copying.
  var params: seq[Tparameter_specification] = @[]

  P(PARAM_VERSION, "Shows the version of the program")
  P(PARAM_HELP, "Shows this help on the commandline", PK_HELP)
  P(PARAM_CONFIG, "Passes a configuration parameter to the command in x=y form",
    PK_STRING, parse_configuration_parameter)
  P(PARAM_PAGINATE, "Pipe all output into less")
  P(COMMAND_ADD, "Add file contents to the index")
  P(COMMAND_MV, "Move or rename a file, a directory, or a symlink")
  P(COMMAND_RM, "Remove files from the working tree and from the index")
  P(COMMAND_DIFF, "Show changes using common diff tools")

  # Cast the result so we don't have to write our own manual assignment proc.
  Tcommandline_results(result) = parse(params)

  if result.options.hasKey(PARAM_VERSION[0]):
    echo "Version 3.1415"
    quit()

  var found_commands: seq[string] = @[]
  for command in COMMANDS:
    if result.options.hasKey(command):
      found_commands.add(command)

  if found_commands.len < 1:
    echo "Please specify a command to work with"
    echo_help(params)
    quit()

  if found_commands.len > 1:
    echo "Please specify a single command to work with, can't use " &
      "$1 at the same time" % [join(found_commands, "|")]
    echo_help(params)
    quit()

  if result.options.hasKey(COMMAND_MV) and result.positional_parameters.len > 2:
    echo "The move command requires a single source and dest params"
    echo_help(params)
    quit()

  if result.positional_parameters.len < 1:
    echo "Pass some paths to work on"
    echo_help(params)
    quit()

  got(PARAM_PAGINATE)

  if result.options.hasKey(PARAM_CONFIG[0]):
    echo "Passing config '$1' to command" % [
      result.options[PARAM_CONFIG[0]].str_val]

  result.command = found_commands[0]


when isMainModule:
  let args = process_commandline()
  echo "Using command '$1'" % [args.command]
  for param in args.positional_parameters:
    echo "Acting on $1" % param.str_val


tester.nim
import argument_parser, strutils, tables

template test_failure(exception, code: expr): stmt =
  try:
    let pos = instantiationInfo()
    discard(code)
    echo "Unit test failure at $1:$2 with '$3'" % [pos.filename,
      $pos.line, astToStr(code)]
    doAssert false, "A test expecting failure succeeded?"
  except exception:
    nil

#template test_success(code: expr) {.immediate.} = echo($code)
template test_success(code: expr) {.immediate.} = discard(code)

template test_in(commandline_parameters, expected_value, attribute: expr):
    expr {.immediate.} =
  ## Returns true if the expected value is found in the commandline parameters.
  ##
  ## Pass the attribute to check in the Tparsed_parameter object.
  var result {.gensym.}: bool = false
  for parameter in commandline_parameters.positional_parameters:
    if parameter.attribute == expected_value:
      result = true
      break
  result

template tp(expected: seq[Tparameter_specification] = @[],
    type_of_positional_parameters = PK_STRING, args: seq[TaintedString] = nil):
      expr =
  # Simple wrapper to avoid changing the last default parameter.
  parse(expected, type_of_positional_parameters, args,
    quit_on_failure = false)

proc test() =
  #echo "\nParsing default system params"
  let
    p1 = new_parameter_specification(single_word = "a", consumes = pkString)
    p2 = new_parameter_specification(double_word = "aasd", consumes = pkString)
    p3 = new_parameter_specification(single_word = "i", consumes = pkInt)
    p4 = new_parameter_specification(single_word = "f", consumes = pkFloat)
    p5 = new_parameter_specification(single_word = "b", consumes = pkBool)
    p6 = new_parameter_specification(single_word = "I", consumes = pkBiggestInt)
    p7 = new_parameter_specification(single_word = "F",
      consumes = pkBiggestFloat)
    all_params = @[p1, p2, p3, p4, p5, p6, p7]

  discard(tp(all_params))

  let args = @["test", "toca me la", "-a", "-wo", "rd", "--aasd", "--s", "ugh"]
  #echo "\nParsing ", join(args, " ")
  var ret = tp(all_params, args = args)
  #echo($ret)
  doAssert ret.options["-a"].strVal == "-wo"
  doAssert ret.options.hasKey("test") == false
  doAssert test_in(ret, "test", str_val)
  doAssert test_in(ret, "--s", str_val) == false
  doAssert test_in(ret, "ugh", str_val)
  test_failure(EInvalidValue, tp(all_params, PK_INT, args))

  # Integer tests.
  test_success(tp(all_params, args = @["int test", "-i", "445"]))
  test_success(tp(all_params, args = @["int test", "-i", "-445"]))
  test_failure(EInvalidValue, tp(all_params, args = @["-i", ""]))
  test_failure(EInvalidValue, tp(all_params, args = @["-i", "0x02"]))
  test_failure(EInvalidValue, tp(all_params, args = @["-i", "fail"]))
  test_failure(EInvalidValue, tp(all_params, args = @["-i", "234.12"]))
  test_failure(EOverflow, tp(all_params, args = @["-i", $high(int) & "0"]))
  test_success(tp(all_params, PK_INT, @["-i", "-445", "2", "3", "4"]))
  test_failure(EInvalidValue,
    tp(all_params, PK_INT, @["-i", "-445", "2", "3", "4.3"]))

  ret = tp(all_params, PK_INT, @["-i", "-445", "2", "3", "4"])
  doAssert ret.options["-i"].int_val == -445
  doAssert test_in(ret, 2, int_val)
  doAssert test_in(ret, 3, int_val)
  doAssert test_in(ret, 4, int_val)
  doAssert test_in(ret, 5, int_val) == false

  # String tests.
  test_success(tp(all_params, args = @["str test", "-a", "word"]))
  test_success(tp(all_params, args = @["str empty test", "-a", ""]))
  test_failure(EInvalidValue, tp(all_params, args = @["str test", "-a"]))

  # Float tests.
  test_success(tp(all_params, args = @["-f", "123.235"]))
  test_failure(EInvalidValue, tp(all_params, args = @["-f", ""]))
  test_failure(EInvalidValue, tp(all_params, args = @["-f", "abracadabra"]))
  test_failure(EInvalidValue, tp(all_params, args = @["-f", "12.34aadd"]))
  ret = tp(all_params, PK_FLOAT, @["-f", "12.23", "89.2", "3.14"])
  doAssert ret.options["-f"].float_val == 12.23
  doAssert test_in(ret, 89.2, float_val)
  doAssert test_in(ret, 3.14, float_val)
  doAssert test_in(ret, 3.1, float_val) == false

  # Boolean tests.
  for param in @["y", "yes", "true", "1", "on", "n", "no", "false", "0", "off"]:
    test_success(tp(all_params, args = @["-b", param]))
  test_failure(EInvalidValue, tp(all_params, args = @["-b", "t"]))
  test_failure(EInvalidValue, tp(all_params, args = @["-b", ""]))
  doAssert test_in(tp(all_params, PK_BOOL, @["y"]), true, bool_val)
  doAssert test_in(tp(all_params, PK_BOOL, @["0"]), false, bool_val)

  # Big integer tests.
  test_success(tp(all_params, args = @["int test", "-I", "445"]))
  test_failure(EInvalidValue, tp(all_params, args = @["-I", ""]))
  test_failure(EInvalidValue, tp(all_params, args = @["-I", "fail"]))
  test_failure(EInvalidValue, tp(all_params, args = @["-I", "234.12"]))
  test_failure(EOverflow, tp(all_params,
    args = @["-I", $high(biggestInt) & "0"]))
  ret = tp(all_params, PK_BIGGEST_INT, @["42", $high(biggestInt)])
  doAssert test_in(ret, 42, big_int_val)
  doAssert test_in(ret, high(biggestInt), big_int_val)
  doAssert test_in(ret, 13, big_int_val) == false

  # Big float tests.
  test_success(tp(all_params, args = @["-F", "123.235"]))
  test_failure(EInvalidValue, tp(all_params, args = @["-F", ""]))
  test_failure(EInvalidValue, tp(all_params, args = @["-F", "abracadabra"]))
  test_failure(EInvalidValue, tp(all_params, args = @["-F", "12.34aadd"]))
  ret = tp(all_params, PK_BIGGEST_FLOAT, @["111.111", "9.01"])
  doAssert test_in(ret, 111.111, big_float_val)
  doAssert test_in(ret, 9.01, big_float_val)
  doAssert test_in(ret, 9.02, big_float_val) == false

  # Using custom procs for transformation of type back to string.
  var c1 = new_parameter_specification(single_word = "i", consumes = PK_INT)
  ret = tp(@[c1], args = @["-i", "42"])
  doAssert ret.options["-i"].int_val == 42
  # Now repeat transforming to int through a custom post-validator.
  c1.custom_validator =
    proc (parameter: string, value: var Tparsed_parameter): string =
      value = new_parsed_parameter(PK_STRING, $value.int_val)
  ret = tp(@[c1], args = @["-i", "42"])
  doAssert ret.options["-i"].str_val == "42"

  # Change the custom proc to reject values lower than 18.
  c1.custom_validator =
    proc (parameter: string, value: var Tparsed_parameter): string =
      let age = value.int_val
      if age < 18:
        result = "Can't accept minors ($1) parsing arguments" % [$age]
      else:
        value = new_parsed_parameter(PK_STRING, "valid_$1" % [$age])
  ret = tp(@[c1], args = @["-i", "42"])
  doAssert ret.options["-i"].str_val == "valid_42"
  test_failure(EInvalidValue, tp(@[c1], args = @["-i", "17"]))

  echo "Tester finished"

when isMainModule:
  test()
